util.c

```diff
@@ -559,6 +559,173 @@ int pad_file(struct image *image, const char *infile,
 	return ret;
 }
 
+/*
+ * Write @size zero bytes at the @offset in @fd. Roughly equivalent to
+ * a single "pwrite(fd, big-all-zero-buffer, size, offset)", except
+ * that we try to use more efficient operations (ftruncate and
+ * fallocate). This only uses methods that do not affect the offset of
+ * fd.
+ */
+static int write_zeroes(int fd, size_t size, off_t offset)
+{
+	struct stat st;
+
+	if (!size)
+		return 0;
+
+	if (fstat(fd, &st) < 0)
+		return -errno;
+
+	if (S_ISREG(st.st_mode)) {
+		if (offset + size > (size_t)st.st_size) {
+			if (ftruncate(fd, offset + size) < 0)
+				return -errno;
+			/*
+			 * The area from st.st_size to offset+size is
+			 * zeroed by this operation. If offset was >=
+			 * st.st_size, we're done.
+			 */
+			if (offset >= st.st_size)
+				return 0;
+			/*
+			 * Otherwise, reduce size accordingly, we only
+			 * need to write zeroes from offset until the
+			 * old end of the file.
+			 */
+			size = st.st_size - offset;
+		}
+		/*
+		 * Maybe this should be guarded by a #ifdef
+		 * HAVE_FALLOCATE. That's easy to add, and the code
+		 * will just automatically fall through to the write
+		 * loop, the same way as if the filesystem doesn't
+		 * support FALLOC_FL_PUNCH_HOLE.
+		 */
+		if (fallocate(fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE,
+			      offset, size) == 0)
+			return 0;
+	}
+
+	/* Not a regular file, or fallocate not applicable. */
+	while (size) {
+		static char buf[4096];
+		size_t now = min(size, sizeof(buf));
+		int r;
+
+		r = pwrite(fd, buf, now, offset);
+		if (r < 0)
+			return -errno;
+		size -= r;
+		offset += r;
+	}
+	return 0;
+}
+
+/*
+ * Insert the image @sub at offset @offset in @image. If @sub is
+ * smaller than @size (including if @sub is NULL), insert 0 bytes for
+ * the remainder. If @sub is larger than @size, only the first @size
+ * bytes of it will be copied (it's up to the caller to ensure this
+ * doesn't happen). This means that after this call, exactly the range
+ * [offset, offset+size) in the output @image have been updated.
+ */
+int insert_image(struct image *image, struct image *sub,
+		 unsigned long long size, unsigned long long offset)
```

Comment:
So, I'm not too happy about adding another function to add stuff to the image. On the one hand, `pad_file()` is a complex beast and I'm reluctant to make it even more complex. On the other hand, most of the complexity was added for the HD image. I think I need to look at the final result some more. Maybe we can do some refactoring and share some code.

Reply:
I agree, but it was simply too hard to modify pad_file to avoid its assumption that one always appends to the current image. Of the three remaining users, the one that uses it to wipe an existing ext[234] header can clearly be replaced directly by an insert_image() with a NULL sub image. That leaves the two uses in flash, but if insert_image learns a fillpattern argument, they could easily be converted as well (so write_zeroes would need to be renamed to write_bytes, but its implementation wouldn't need to change very much).

I don't very much like that magic ->last_offset which pad_file maintains just for itself, nor the magic of "pass MODE_OVERWRITE on the first call, then MODE_APPEND", and a better model is that the callers know what they want to insert where. So if those two flash users could get converted, I don't think there's gonna be much of a net line count addition.

Reply:
That code evolved... :-). Originally, the code just added at the end of the file. But the I wanted to write to block devices as well. So `last_offset` basically simulates the current "end"... We don't need all that if the used explicitly specifies the offset.

That reminds me that I need to check if `insert_image()` works correctly for block devices as well. Unfortunately that's not something that can be tested in the test-suite the way it works right now.

Replacing `pad_image()` everywhere sound like a good plan. I'll be happy to take patches for it, but I won't make it a requirement for this PR go get merged.

Reply:
Yeah, I figured it had grown organically, but I decided that this was the time to stop trying to make it handle even more cases (because a MODE_INSERT would be fundamentally incompatible with maintaining the ->last_offset, and the very name "pad_file" implies working at the end).

Re block devices, I haven't tested (yes, it's annoyingly difficult to do, especially without root), but I think the model with the caller giving the offset should in general make it "just work", the same way that it already works for insert_data().

And I'll be happy to write the patches to remove the last users of pad_file, but not as part of this series. That will then involve changing the prototype of insert_file and thus update the users I introduce here, but that's mostly mechanical.